### PR TITLE
Daemonize helper process on Darwin

### PR DESCRIPTION
**Pull request by mrmacete**
In this way there’s no need to install a `SIGCHLD` handler which could conflict with others already installed in the process, causing issues for example with NodeJS and its `child_process` logic.

### CODE REVIEW

File: src/darwin/frida-helper-process.vala [hunk 1]
```diff
@@ -267,9 +264,10 @@ namespace Frida {
 				var handshake_port = new HandshakePort.local (service_name);
 
 				string[] argv = { get_resource_store ().helper.path, service_name };
-				pending_process = new Subprocess.newv (argv, SubprocessFlags.INHERIT_FDS);
 
-				var peer_pid = (uint) uint64.parse (pending_process.get_identifier ());
+				GLib.SpawnFlags flags = GLib.SpawnFlags.LEAVE_DESCRIPTORS_OPEN | /*GLib.SpawnFlags.CLOEXEC_PIPES*/ 256;
```
Comment (mrmacete) [inline src/darwin/frida-helper-process.vala:268]: I don't know why, but apparently `GLib.SpawnFlags.CLOEXEC_PIPES` doesn't seem to be exposed (even if it's documented) and i get an error like:
```
../../../frida-core/src/darwin/frida-helper-process.vala:266.70-266.98: error: The name `CLOEXEC_PIPES' does not exist in the context of `GLib.SpawnFlags' (glib-2.0)
				GLib.SpawnFlags flags = GLib.SpawnFlags.LEAVE_DESCRIPTORS_OPEN | GLib.SpawnFlags.CLOEXEC_PIPES;
				                                                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

Comment (mrmacete) [reply src/darwin/frida-helper-process.vala:268]: i'm not even sure we really need it, i kept it for compatibility with what SubProcess was already doing here.

Comment (oleavr) [inline src/darwin/frida-helper-process.vala:268]: Makes sense!  Yeah this looks like a vapi binding bug.

```suggestion
				GLib.SpawnFlags flags = GLib.SpawnFlags.LEAVE_DESCRIPTORS_OPEN | /* GLib.SpawnFlags.CLOEXEC_PIPES */ 256;
```

File: src/darwin/frida-helper-process.vala [hunk 2]
```diff
@@ -16,8 +13,8 @@ namespace Frida {
 
 		private ResourceStore _resource_store;
 
+		private uint process_pid;
```
Comment (oleavr) [inline src/darwin/frida-helper-process.vala:16]: Let's move this down to where `process` was.

File: src/darwin/frida-helper-process.vala [hunk 3]
```diff
@@ -255,7 +252,7 @@ namespace Frida {
 			}
 			obtain_request = new Promise<DarwinRemoteHelper> ();
 
-			Subprocess pending_process = null;
+			uint peer_pid = 0;
```
Comment (oleavr) [inline src/darwin/frida-helper-process.vala:255]:
```suggestion
			uint pending_pid = 0;
```

(To make it clear it belongs with the other ones.)

File: src/darwin/frida-helper-process.vala [hunk 4]
```diff
@@ -302,8 +300,9 @@ namespace Frida {
 				obtain_request.resolve (proxy);
 				return proxy;
 			} else {
-				if (pending_process != null)
-					pending_process.force_exit ();
+				process_pid = 0;
```
Comment (oleavr) [inline src/darwin/frida-helper-process.vala:303]: Shouldn't need to clear it here as we haven't set it.

**Comment by mrmacete** [comment]
I couldn't reproduce the error in the same way in Linux. Since it's using the same APIs it's possible that the problem is there as well, just i'm not sure how to reproduce it systematically.

I could do the same tweak also there but i'm a bit blind for validating it.

**Comment by oleavr** [comment]
> I could do the same tweak also there but i'm a bit blind for validating it.

That would be great! I've seen random test failures on Linux that look a lot like the same underlying issue. It's very racy though, so not easy to verify.

**Pull request merged**